### ISSUE TITLE
mgmtCmd EXECUTE response?

### ISSUE THREAD

**Issue by Spongman** [oneM2M/Issues-Questions#4]
in TS-0001 Table 10.2.8.18-1: \<mgmtCmd\> EXECUTE it says "It shall also provide in the response the URL of the created <execInstance> resource"
- what is the mechanism for this?
- presumably this can be multiple if a \<group\> is specified?

**Comment by ankraft** [comment]
I thing the full text reads

    It shall also provide in the response the URL of the created \<execInstance> resource.

This is similar to other async or non-blocking requests: The CSE creates a resource that holds the original request, has the state of the request and will later also hold the response of it. The request originator needs to know the URI of the created instance in order to check its status and retrieve a response later. That is this URI. 

> presumably this can be multiple if a <group> is specified?

No, I don't think that \<group> is involved here. And even if it would be, then the response to a` <group>/fopt` would be a single aggregated response.

**Comment by Spongman** [comment]
but this is not in response to a CREATE request, it's an UPDATE on the mgmtCmd's execEnable:

> The Execute procedure shall be used by an Originator in order to trigger execution of a specific management command on a managed entity,  by employing an UPDATE method to the execEnable attribute of an existing <mgmtCmd> resource on the Hosting CSE.

also, wrt. groups:

> if the execTarget attribute of the addressed <mgmtCmd> addresses a group, the Hosting CSE shall create corresponding <execInstance> resources for each target in the group and provide the corresponding URLs in the response

8.1.3 specifies the form of the response to an UPDATE request. where are these new execTarget URLs suppsoed to go?

**Comment by ankraft** [comment]
You are right. I remembered a different procedure.

In TS-0004, clause 7.4.16.2.3.2 Update (Execute), the procedure is explained in detail. Step 3:

> 3) The Receiver shall copy the following attributes from <mgmtCmd> to each <execInstance> created: execMode, execFrequency, execDelay, execNumber, and execReqArgs. The execStatus of <execInstance> is set to INITIATED. **The Receiver shall set the execTarget attribute of each <execInstance> sub-resource to the URI of each target <node> resource.**

**Comment by Spongman** [comment]
sure, but those copied attributes are placed in the new `execTarget` resource. the response to the UPDATE request, however, only contains attributes of the resource being updated - the `mgmtCmd`. again: where are the URLs of the new `execTarget` resources returned?

**Comment by Spongman** [comment]
also, how does one EXECUTE a single `mgmtCmd` at two different times to two different targets? is it possible to perform two different UPDATE requests that each update _both_ the `execTarget` AND `execEnable` attributes simultaneously?

**Comment by ankraft** [comment]
> sure, but those copied attributes are placed in the new `execTarget` resource. the response to the UPDATE request, however, only contains attributes of the resource being updated - the `mgmtCmd`. again: where are the URLs of the new `execTarget` resources returned?

I have to say that this seems to be missing for now. One can do a DISCOVERY request to find the \<execInstance> resource, but this is not convenient at all. I agree, there should be a RO attribute that is a list of all the \<execInstance> resources for a particular UPDATE (execution) request.

**Comment by ankraft** [comment]
> also, how does one EXECUTE a single mgmtCmd at two different times to two different targets? is it possible to perform two different UPDATE requests that each update both the execTarget AND execEnable attributes simultaneously?

I think this should be possible. A new EXECUTE should create new \<execInstance> resources for that request.

**Comment by Spongman** [comment]
yeah, doing a DISCOVERY to find out which `execInstance` was the one created by the previous UPDATE is obviously prone to sequencing errors if there are multiple simultaneous UPDATE calls in-flight.
